Source/Core/DolphinQt/BbaServerWindow.cpp

```diff
@@ -0,0 +1,100 @@
+// Copyright 2019 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include "BbaServerWindow.h"
+
+#include <QVBoxLayout>
+#include <QHBoxLayout>
+#include <QLabel>
+#include <QLineEdit>
+#include <QSpinBox>
+#include <QPushButton>
+#include <QPlainTextEdit>
+#include <QDialogButtonBox>
+#include <QMessageBox>
+#include <QTime>
+#include <QScrollBar>
+
+#include <limits>
+
+BbaServerWindow::BbaServerWindow(QWidget *parent) :
+  QDialog(parent),
+  m_server(this)
+{
+  auto vboxLayout = new QVBoxLayout(this);
+
+  {
+    auto hboxLayout = new QHBoxLayout;
+
+    auto hostAddrLabel = new QLabel(tr("Host address:"), this);
+    hboxLayout->addWidget(hostAddrLabel);
+
+    m_hostAddr = new QLineEdit(this);
+    hostAddrLabel->setBuddy(m_hostAddr);
+    m_hostAddr->setPlaceholderText(tr("Leave empty for Any"));
+    hboxLayout->addWidget(m_hostAddr);
+
+    auto portLabel = new QLabel(tr("Port:"), this);
+    hboxLayout->addWidget(portLabel);
+
+    m_port = new QSpinBox(this);
+    portLabel->setBuddy(m_port);
+    m_port->setRange(std::numeric_limits<quint16>::min(), std::numeric_limits<quint16>::max());
+    m_port->setValue(50000);
+    hboxLayout->addWidget(m_port);
+
+    m_toggle = new QPushButton(this);
+    connect(m_toggle, &QAbstractButton::pressed, this, &BbaServerWindow::toggle);
+    hboxLayout->addWidget(m_toggle);
+
+    vboxLayout->addLayout(hboxLayout);
+  }
+
+  m_logOutput = new QPlainTextEdit(this);
+  m_logOutput->setLineWrapMode(QPlainTextEdit::NoWrap);
+  m_logOutput->setReadOnly(true);
+  vboxLayout->addWidget(m_logOutput, 1);
+
+  {
+    auto buttonBox = new QDialogButtonBox(QDialogButtonBox::Close, this);
+    connect(buttonBox, &QDialogButtonBox::rejected, this, &QDialog::reject);
+    vboxLayout->addWidget(buttonBox, 1);
+  }
+
+  setLayout(vboxLayout);
+
+  connect(&m_server, &BbaServer::information, this, &BbaServerWindow::logOutput);
+
+  update();
+}
+
+void BbaServerWindow::toggle()
+{
+  if (m_server.isListening())
+    m_server.close();
+  else
+  {
+    if (!m_server.listen(m_hostAddr->text(), m_port->value()))
+      QMessageBox::warning(this, tr("Could not start listening!"), QStringLiteral("%0\n\n%1").arg(tr("Could not start listening!"),m_server.errorString()));
```

Comment:
You'll likely want to make the format string translatable, otherwise this can result in incorrect ordering for languages that have different ways of ordering text.